runtime/error.c

```diff
@@ -1,9 +1,116 @@
+#include <string.h>
+#include <stdlib.h>
+#include <stdio.h>
+
 #include "balrt.h"
+#include "third-party/libbacktrace/backtrace.h"
+
+#define MIN_FRAME_COUNT 2
+#define SKIP_FROM_END 1 // This number counts from the last stack frame
+#define THREAD 0
+
+typedef struct {
+    uint32_t nFrames;
+    uint32_t capacityOfFrames;
+    uint32_t nBytes;
+    uint32_t capacityOfBytes;
+    Frame frames[];
+} FramesInfo;
+
+void on_error(void *vdata, const char *msg, int errnum) { printf("Error : %s\n", msg); }
+
+int on_frame(void *data, uintptr_t pc, const char *filename, int lineno, const char *function) {
+    // This check is valid if there is a posibility not to have debug info
+    if (filename == NULL || function == NULL)
+    {
+        return 0;
+    }
+
+    FramesInfo *framesInfo = (FramesInfo *)data;
+    Frame *frames = framesInfo->frames;
+    uint32_t nFrames = framesInfo->nFrames;
+    uint32_t capacityOfFrames = framesInfo->capacityOfFrames;
+    uint32_t nBytes = framesInfo->nBytes;
+    uint32_t capacityOfBytes = framesInfo->capacityOfBytes;
+
+    uint64_t filenameLen = strlen(filename);
+    uint64_t funcLen = strlen(function);
+
+    if (nFrames == capacityOfFrames || (nBytes + filenameLen + funcLen + 2) >= capacityOfBytes)
+    {
+        // The current total memory size of bytes array and frames array is multiplied by two.
+        // New allocation should be devided equally between bytes array and frames array.
+        uint32_t newSize = sizeof(FramesInfo) + ((capacityOfFrames * sizeof(Frame) + capacityOfBytes) << 1);
+        framesInfo = realloc(framesInfo, newSize);
+
+        uint32_t newCapacityOfFrames = capacityOfFrames << 1;
+        memcpy(frames + newCapacityOfFrames, frames + capacityOfFrames, nBytes);
+        framesInfo->capacityOfFrames = newCapacityOfFrames;
+        framesInfo->capacityOfBytes = capacityOfBytes << 1;
+        capacityOfFrames = newCapacityOfFrames;
+    }
+
+    Frame *frame = frames + nFrames;
+    frame->pc = pc;
+    frame->lineno = lineno;
+
+    char *bytes = (char *)(frames + capacityOfFrames) + nBytes;
+
+    memcpy(bytes, filename, filenameLen);
+    frame->fileOffset = nBytes;
+    bytes = bytes + filenameLen;
+    nBytes = nBytes + filenameLen;
+    *bytes = '\0';
+    bytes++;
+    nBytes++;
+
+    memcpy(bytes, function, funcLen);
+    frame->functionOffset = nBytes;
+    bytes = bytes + funcLen;
+    nBytes = nBytes + funcLen;
+    *bytes = '\0';
+    bytes++;
+    nBytes++;
+
+    framesInfo->nBytes = nBytes;
+    framesInfo->nFrames = nFrames + 1;
+
+    return 0;
+}
+
+FramesInfo *getFramesInfo() {
+    struct backtrace_state *state = backtrace_create_state(NULL, THREAD, on_error, NULL);
+
+    // The memory size of bytes array and frames array are equal
+    // Therefore minimum frame size is multiplied by 2
+    FramesInfo *framesInfo = malloc(sizeof(FramesInfo) + MIN_FRAME_COUNT * sizeof(Frame) * 2);
```

Comment:
Completely unacceptable to not check return value of malloc.

Reply:
Fixed in all places.